File: HOT100/5.普通数组/238.除自身以外数组的乘积.cpp
/*
 * @lc app=leetcode.cn id=238 lang=cpp
 *
 * [238] 除自身以外数组的乘积
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> suf(nums.size());
        int pre = nums[0];
        suf[nums.size() - 1] = nums[nums.size() - 1];
        for (int i = nums.size() - 2; i > 0; i -- )
        {
            suf[i] = suf[i + 1] * nums[i];
        }
        suf[0] = suf[1];
        for (int i = 1; i < nums.size() - 1; i ++ )
        {
            suf[i] = pre * suf[i + 1];
            pre *= nums[i];
        }
        suf[nums.size() - 1] = pre;
        return suf;
    }
};
// @lc code=end

class Solution {
    public:
        vector<int> productExceptSelf(vector<int>& nums) {
            vector<int> pre(nums.size()), suf(nums.size());
            vector<int> res(nums.size());
            pre[0] = nums[0], suf[nums.size() - 1] = nums[nums.size() - 1];
            for (int i = 1; i < nums.size(); i ++ )
            {
                pre[i] = pre[i - 1] * nums[i];
            }
            for (int i = nums.size() - 2; i >= 0; i -- )
            {
                suf[i] = suf[i + 1] * nums[i];
            }
            res[0] = suf[1], res[nums.size() - 1] = pre[nums.size() - 2];
            for (int i = 1; i < nums.size() - 1; i ++ )
            {
                res[i] = pre[i - 1] * suf[i + 1];
            }
            return res;
        }
    };

File: 面经150/数组&字符串/238.除自身以外数组的乘积.py
#
# @lc app=leetcode.cn id=238 lang=python3
#
# [238] 除自身以外数组的乘积
#
from mytools import *
# @lc code=start
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = nums[:]
        n = len(nums)
        for i in range(n - 2, -1, -1): nums[i] *= nums[i + 1]
        pre = 1
        for i in range(0, n):
            tmp = res[i]
            res[i] = pre * (nums[i + 1] if i + 1 < n else 1)
            pre *= tmp
        return res
# @lc code=end

